### PR TITLE
[JIT] Allow new temp assignment whose dstTyp is TYP_I_IMPL and valTyp is TYP_BYREF

### LINKED ISSUE #78944: Assertion failed '!"Incompatible types for gtNewTempAssign"'
win-x86 checked random jitstress

https://dev.azure.com/dnceng-public/public/_build/results?buildId=94457&view=ms.vss-test-web.build-test-results-tab&runId=1991328&paneView=debug&resultId=119215

Failures:

JIT\Directed\coverage\importer\byrefsubbyref1\byrefsubbyref1.cmd
JIT\Directed\coverage\importer\Desktop\byrefsubbyref1_il_r\byrefsubbyref1_il_r.cmd
JIT\Directed\coverage\importer\Desktop\byrefsubbyref1_il_d\byrefsubbyref1_il_d.cmd

```
    JIT\Directed\coverage\importer\Desktop\byrefsubbyref1_il_r\byrefsubbyref1_il_r.cmd [FAIL]
      
      Assert failure(PID 1364 [0x00000554], Thread: 7036 [0x1b7c]): Assertion failed '!"Incompatible types for gtNewTempAssign"' in 'a:byrefsubi4(byref,int):int' during 'Morph - Global' (IL size 10; hash 0x72f297d0; FullOpts)
      
          File: D:\a\_work\1\s\src\coreclr\jit\gentree.cpp Line: 15705
          Image: C:\h\w\B23E091A\p\corerun.exe
      
      
      Return code:      1
      Raw output file:      C:\h\w\B23E091A\w\A9E30972\uploads\Reports\JIT.Directed\coverage\importer\Desktop\byrefsubbyref1_il_r\byrefsubbyref1_il_r.output.txt
      Raw output:
      BEGIN EXECUTION
       "C:\h\w\B23E091A\p\corerun.exe" -p "System.Reflection.Metadata.MetadataUpdater.IsSupported=false"  byrefsubbyref1_il_r.dll 
      -4
                     [000002] -----+-----                         *  SUB       byref 
                     [000000] -----+-----                         +--*  LCL_VAR   byref  V00 arg0         
                     [000001] -----+-----                         \--*  LCL_VAR   int    V01 arg1         
      Expected: 100
      Actual: -1073740286
      END EXECUTION - FAILED
```

with:
```
set DOTNET_TieredCompilation=0
set DOTNET_JitStress=2b1
```

@dotnet/jit-contrib

### PR DESCRIPTION
**Pull request by TIHan**
**Description**

Should resolve: https://github.com/dotnet/runtime/issues/78944

I have two other PRs that tried to resolve this issue as well:

- https://github.com/dotnet/runtime/pull/78999

- https://github.com/dotnet/runtime/pull/79113

When creating a temp assignment, we would assert if the new temp assignment looked like this `ASG(VAR{i_impl}, NODE{byref})`. This PR now allows these types when creating a new temp assignment, assuming that `ASG(VAR{i_impl}, NODE{byref})` is considered legal IR.

I believe that IR is legal, because consider this example:
```csharp
struct ctest { }
unsafe static IntPtr byrefsubi4(ref ctest V_1, int V_2)
{
    return (nint)Unsafe.AsPointer(ref Unsafe.SubtractByteOffset(ref V_1, V_2));
}
```

After global morph, the IR looks like this:
```
               [000005] -A---+-----                         *  RETURN    long  
               [000010] -A---+-----                         \--*  COMMA     long  
               [000007] -A---+-----                            +--*  ASG       long  
               [000006] D----+-N---                            |  +--*  LCL_VAR   long   V03 tmp1         
               [000003] -----+-----                            |  \--*  SUB       byref 
               [000000] -----+-----                            |     +--*  LCL_VAR   byref  V00 arg0         
               [000002] -----+-----                            |     \--*  CAST      long <- int
               [000001] -----+-----                            |        \--*  LCL_VAR   int    V01 arg1         
               [000008] -----+-----                            \--*  LCL_VAR   long   V03 tmp1
```
Notice the `ASG(VAR{long}, NODE{byref})` - so it seems like we allow this today.
In this case, the above IR's `tmp1` is created when creating a temp for cast-away gc info:
```cpp
    else if (varTypeIsGC(srcType) != varTypeIsGC(dstType))
    {
        // We are casting away GC information.  we would like to just
        // change the type to int, however this gives the emitter fits because
        // it believes the variable is a GC variable at the beginning of the
        // instruction group, but is not turned non-gc by the code generator
        // we fix this by copying the GC pointer to a non-gc pointer temp.
        noway_assert(!varTypeIsGC(dstType) && "How can we have a cast to a GCRef here?");

        // We generate an assignment to an int and then do the cast from an int. With this we avoid
        // the gc problem and we allow casts to bytes, longs,  etc...
        unsigned lclNum = lvaGrabTemp(true DEBUGARG("Cast away GC"));
        oper->gtType    = TYP_I_IMPL;
        GenTree* asg    = gtNewTempAssign(lclNum, oper);
        oper->gtType    = srcType;
```

The reason why this case does not assert is because we change oper type right before `gtNewTempAssign` which is `oper->gtType = TYP_I_IMPL`. And then right after it, we set it back so its `srcType`. So that explains why the assert didn't happen here.

### CODE REVIEW

File: src/coreclr/jit/gentree.cpp [hunk 1]
```diff
@@ -15683,6 +15683,10 @@ GenTree* Compiler::gtNewTempAssign(
         {
             ok = true;
         }
+        else if ((dstTyp == TYP_I_IMPL) && varTypeIsGC(valTyp))
```
Comment (SingleAccretion) [inline src/coreclr/jit/gentree.cpp:15686]: > `varTypeIsGC`

This includes `TYP_REF` as well, do we need that to fix the asserts?

Comment (TIHan) [reply src/coreclr/jit/gentree.cpp:15686]: I guess not? I can just do it by BYREF only.

**Comment by TIHan** [comment]
@dotnet/jit-contrib @jakobbotsch - I am marking this as ready for review as I think this is the proper fix for the issue if we only consider what is valid IR, which I now believe that `ASG(VAR{i_impl}, NODE{byref})` is because we have this IR shape occurring today.

**Comment by BruceForstall** [comment]
1. Why are we seeing this assertion now? What is the root cause of the regression? Why didn't we see it in PR testing?
2. You give an example of code above that goes to pains to avoid triggering the assert. Why aren't we going through this same code path for this test case? What code path are we going through, and should it have similar code to the above (i.e., set type, call gtNewTempAssign, set type again)?

**Comment by TIHan** [comment]
I tested `release/7.0` branch and was not able to reproduce it with the same JitStress mode. Am looking further.

**Comment by TIHan** [comment]
I found the cause: https://github.com/dotnet/runtime/pull/77655

We removed this check from `gtNewTempAssign`:
```diff
-        else if (!varTypeIsGC(dstTyp) && (genTypeSize(valTyp) == genTypeSize(dstTyp)))
-        {
-            // We can have assignments that require a change of register file, e.g. for arguments
-            // and call returns. Lowering and Codegen will handle these.
-            ok = true;
-        }
```

--

This makes me more confident that the change in this PR resolves the issue and is correct.

**Review by BruceForstall — APPROVED** [review]
Thanks for tracking down the reason.

**Pull request merged**